### DISCUSSION TITLE
[NEW PASS] AI-Driven Resource-Aware Router (AIRA)

### DISCUSSION THREAD

**Discussion by poig** [unitaryfoundation/ucc#410] [New Compiler Pass]
## Abstract
This proposal introduces the AI-Driven Resource-Aware Router (AIRA), a novel transformation pass for quantum circuit compilation that leverages a deep learning model to significantly outperform traditional heuristic-based approaches. While standard transpilers like Qiskit's often focus on minimizing SWAP gate insertions based on static connectivity, they can produce circuits with excessive depth and gate counts, increasing their susceptibility to decoherence and operational errors on NISQ devices. 

My approach formulates qubit routing as a predictive, AI-guided search problem. The core of AIRA is a Graph Transformer model, pre-trained on a large and diverse dataset of quantum circuits. This model learns the complex, non-local correlations between circuit structure, hardware noise characteristics, and the final, compiled circuit's quality.

TL;DR: I let AI speak quantum language with original transformer mode! where it shows a very good training calibration and potential compare to the best passes. 
Sadly I am still in exam period, I can't go further than that

The compilation pipeline operates as follows:

- Just-in-Time (JIT) Profiling: Inspired by the work of Wilson et al., the pass first builds a DeviceNoiseProfile using the most recent hardware calibration data, ensuring decisions are based on the current state of the device.
- Heuristic Candidate Generation: Rather than having the AI search an intractably large space, a fast, SABRE-like heuristic (inspired by Niu et al.) proposes a small set of high-quality candidate SWAP sequences to resolve routing conflicts.
- AI-Powered Selection: The Transformer model evaluates each candidate sequence, predicting the final circuit's fidelity, depth, and gate count. It selects the sequence that yields the best outcome on a multi-objective cost function.

This hybrid approach differs significantly from existing methods. Unlike IBM's AI passes, which often focus on specific synthesis tasks (e.g., decomposing unitaries), AIRA is a general-purpose router designed to optimize any circuit for a given hardware backend. By combining the speed of heuristics with the global insight of a Transformer, AIRA produces circuits that are demonstrably more compact and efficient, making them more viable for execution on today's noisy quantum hardware.

notebook: https://github.com/poig/ucc/blob/noise_aware-passes/ucc/noise_aware/NoiseAwareRouting.ipynb
fork: https://github.com/poig/ucc/tree/noise_aware-passes

## Diagram of Results
The following benchmark compares AIRA against Qiskit's optimization_level=3 and the default UCC compiler on QFT circuits of increasing complexity.
![image](https://github.com/user-attachments/assets/bd85a5c7-1fd6-4527-89c3-973cf7e2184f)

## What circuit types will this improve?
This pass is expected to provide the most significant improvements for circuits that are highly-entangled and structured, requiring non-trivial routing on devices with limited connectivity. Our benchmarks on Quantum Fourier Transform (QFT) circuits confirm this. The benefits will be most pronounced on circuits where the logical connectivity does not align well with the hardware topology, forcing the compiler to make many strategic SWAP decisions.

## Which UCC metrics will it impact?
Based on our benchmarks, the AIRA pass will have a dramatic and positive impact on the following key metrics:
### **Comprehensive Performance Analysis**

The benchmark results, displayed across three key metrics, demonstrate the superior performance of our AI Router compared to both the default UCC compiler and the highly-optimized Qiskit (Level 3) transpiler.

1.  **Reliability (Predicted Fidelity): Higher is Better**
    The top plot shows that for circuits up to 8 qubits, **our AI Router achieves a significantly higher predicted fidelity** than both competing methods. For instance, at 4 qubits, our router's fidelity is over 0.5, while the competition is below 0.3. As circuit complexity increases, the fidelity of all methods converges toward zero, which is expected due to the exponential difficulty of simulating large, noiseless QFT circuits. However, in the feasible regime, our router is the clear winner on this metric.

2.  **Circuit Compactness: Depth and Gate Count are Lower is Better**
    The bottom two plots reveal the core strength of our approach.
    *   **Circuit Depth:** Our AI Router produces circuits with a drastically lower and nearly constant depth, while the depth of circuits from the other compilers grows rapidly.
    *   **Two-Qubit Gate Count:** Our router consistently requires the fewest two-qubit gates, with its advantage growing as circuit complexity increases. For the largest circuits tested, it uses less than half the number of gates required by Qiskit.

### **Discussion: The True Meaning of "Optimal"**

A naive interpretation of the "Reliability" plot might be to only focus on the predicted fidelity. However, this metric is fundamentally incomplete as it primarily accounts for individual gate errors and **does not factor in decoherence over time (T1/T2 decay)**.

This is where our AI Router's advantage in compactness becomes paramount. A circuit that is 5-10x shallower, as shown in the "Circuit Depth" plot, executes in a fraction of the time on real hardware. This dramatic reduction in execution time means the qubits are far less susceptible to decoherence.

Therefore, we can strongly hypothesize that the already superior fidelity shown in the top plot is an **underestimation of our AI Router's true performance**. On real hardware, where time-based noise is a critical factor, the compact circuits produced by our router would yield significantly more accurate results than the deeper, more complex circuits from the other compilers. Our pass optimizes for the metrics (depth and CNOTs) that are the best proxies for success on real NISQ devices.

References:
E. Wilson, S. Singh, & F. Mueller, "Just-in-time Quantum Circuit Transpilation Reduces Noise," IEEE International Conference on Quantum Computing and Engineering (QCE), 2020.
S. Niu, A. Suau, G. Staffelbach, & A. Todri-Sanial, "A Hardware-Aware Heuristic for the Qubit Mapping Problem in the NISQ Era," IEEE Transactions on Quantum Engineering, 2020.

Here are my ucc-bench result, please check out my readme file over here for full detail: https://github.com/poig/ucc/blob/noise_aware-passes/ucc/noise_aware/ml_model/readme.md

training data can be found here: https://drive.google.com/file/d/1Gd6GkOE18tGs9DLaeHz883JDbsDqEs2h/view?usp=drive_link

**IMPORTANT disclaimer**: please remember that due to computation limitation at data generation, I only generate up to `10k 4-20 qubit` circuit, generated with `ibm_fake_washington` backend data, and I didn't apply any advantage technique yet to calibrate or compress the model to the best, so its just for prototype
```
uv run ucc-bench C:\Users\junli\ucc-bench\benchmarks\layout_benchmarks.toml
```
There plot are generated with the csv at the bottom of the juptyer notebooks
![image](https://github.com/user-attachments/assets/dd7f9db7-af8e-46c9-b0b2-04288dd355c4)
![image](https://github.com/user-attachments/assets/1ca45443-a234-4c92-aa6f-6bb2e112642a)
![image](https://github.com/user-attachments/assets/11ac4b0b-5831-4a53-880a-f054ffa01fd1)
![image](https://github.com/user-attachments/assets/5a037b1c-4bff-4d0c-a22f-39749158fd6a)

data:
```
csv_data = """
compiler,benchmark_id,target_device_id,raw_multiq_gates,compile_time_ms,compiled_multiq_gates

ucc,qaoa,ibm_fake_washington,96,1373.4290,197
ucc,qv,ibm_fake_washington,150,1586.4146,318
ucc,qft,ibm_fake_washington,105,1435.3986,209
ucc,square_heisenberg,ibm_fake_washington,144,903.8763,54
ucc,qcnn,ibm_fake_washington,32,536.7093,49
ucc,qaoa,ibm_fake_brooklyn,96,722.5264,196
ucc,qv,ibm_fake_brooklyn,150,1351.2112,294
ucc,qft,ibm_fake_brooklyn,105,1003.9889,215
ucc,square_heisenberg,ibm_fake_brooklyn,144,682.9114,54
ucc,qcnn,ibm_fake_brooklyn,32,501.1718,52

qiskit-default,qaoa,ibm_fake_washington,96,853.9519,201
qiskit-default,qv,ibm_fake_washington,150,1642.9988,300
qiskit-default,qft,ibm_fake_washington,105,977.5887,208
qiskit-default,square_heisenberg,ibm_fake_washington,144,825.2865,57
qiskit-default,qcnn,ibm_fake_washington,32,504.7062,51
qiskit-default,qaoa,ibm_fake_brooklyn,96,808.6774,194
qiskit-default,qv,ibm_fake_brooklyn,150,1252.0317,306
qiskit-default,qft,ibm_fake_brooklyn,105,926.5525,215
qiskit-default,square_heisenberg,ibm_fake_brooklyn,144,459.7529,57
qiskit-default,qcnn,ibm_fake_brooklyn,32,314.3586,49

ucc-ml-router-128b-8h-6l512,qaoa,ibm_fake_washington,96,1417.3103,197
ucc-ml-router-128b-8h-6l512,qv,ibm_fake_washington,150,1297.4725,318
ucc-ml-router-128b-8h-6l512,qft,ibm_fake_washington,105,1211.0425,209
ucc-ml-router-128b-8h-6l512,square_heisenberg,ibm_fake_washington,144,1083.5276,54
ucc-ml-router-128b-8h-6l512,qcnn,ibm_fake_washington,32,798.2208,49
ucc-ml-router-128b-8h-6l512,qaoa,ibm_fake_brooklyn,96,1358.5344,196
ucc-ml-router-128b-8h-6l512,qv,ibm_fake_brooklyn,150,1309.9864,294
ucc-ml-router-128b-8h-6l512,qft,ibm_fake_brooklyn,105,1033.5248,215
ucc-ml-router-128b-8h-6l512,square_heisenberg,ibm_fake_brooklyn,144,457.3739,54
ucc-ml-router-128b-8h-6l512,qcnn,ibm_fake_brooklyn,32,454.3716,52

ucc-ml-router-256b-8h-8l-1024,qaoa,ibm_fake_washington,96,1425.1702,197
ucc-ml-router-256b-8h-8l-1024,qv,ibm_fake_washington,150,1667.7979,318
ucc-ml-router-256b-8h-8l-1024,qft,ibm_fake_washington,105,1413.7954,209
ucc-ml-router-256b-8h-8l-1024,square_heisenberg,ibm_fake_washington,144,1040.8131,54
ucc-ml-router-256b-8h-8l-1024,qcnn,ibm_fake_washington,32,426.5308,49
ucc-ml-router-256b-8h-8l-1024,qaoa,ibm_fake_brooklyn,96,887.5637,196
ucc-ml-router-256b-8h-8l-1024,qv,ibm_fake_brooklyn,150,1495.5070,294
ucc-ml-router-256b-8h-8l-1024,qft,ibm_fake_brooklyn,105,1167.0798,215
ucc-ml-router-256b-8h-8l-1024,square_heisenberg,ibm_fake_brooklyn,144,766.9933,54
ucc-ml-router-256b-8h-8l-1024,qcnn,ibm_fake_brooklyn,32,324.5032,52
"""
```

with `uv run python .github/scripts/plot_avg_by_time_benchmark.py results ucc-benchmarks-8-core-U22.04`

![image](https://github.com/user-attachments/assets/e7f61f07-35ac-41a4-8508-031237315a89)

with `uv run python .github/scripts/plot_latest_benchmark.py results ucc-benchmarks-8-core-U22.04`
![image](https://github.com/user-attachments/assets/7769c285-8918-44bc-b704-3554d725c8d0)

### Develop your compiler pass

Once the maintainers have given you the go-ahead, you can work on the next sections:

3. Create a [fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) of UCC to develop in: [link your fork here]  
    **Hint:** We recommending [syncing your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) with the main branch of UCC to stay up to date with changes. 
5. Implement and Validate a Prototype of the Pass:  
    A Jupyter notebook or a small script is sufficient for the prototype: [link prototype script/notebook here].  
    **Important:** Make sure your compiler pass works as you expect on the circuits you defined in step 2a.

6. Implement the New Pass in the UCC Codebase:  
Documentation to guide you through this process is available in the [user guide](https://ucc.readthedocs.io/en/latest/user_guide.html). For more detailed information and examples, refer to the [Qiskit documentation](https://docs.quantum.ibm.com/guides/custom-transpiler-pass).

### Benchmark performance

UCC benchmarks live in the separate repo called [ucc-bench](https://github.com/unitaryfoundation/ucc-bench). We have a [suite of quantum circuits](https://github.com/unitaryfoundation/ucc-bench/tree/main/benchmarks) that we regularly benchmark UCC and other popular quantum compilers on. Several of the key metrics we track are:

- Compiling alone:
  - 2-qubit gate count after circuit compilation (lower is better)
  - Total runtime of compilation (lower is better)

- Simulating compiled circuits:
  - Relative errors in simulated expectation values for a defined set of observables (lower is better)

Your new pass should improve one or more of these metrics on our existing benchmark suite.  

> **Note:** If you are introducing a technique whose effect on performance is not currently measurable by our [benchmark suite](https://github.com/unitaryfoundation/ucc-bench/tree/main/benchmarks) (e.g. a new qubit mapping pass that performs very well on sparsely connected qubit layouts), let us know and we can work with you to add the metric into [ucc-bench](github.com/unitaryfoundation/ucc-bench).

#### When you are ready to run benchmarks...  

6. Open a [pull request (PR) to merge your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork) of UCC in main (mark it as a Draft unless you are ready for final review), ping a UCC maintainer, and we will trigger the benchmarking suite to run. Alternately, you can also run the benchmarks locally by comparing your UCC version to main, as explained in the [ucc-bench README](https://github.com/unitaryfoundation/ucc-bench#). 
---

### Useful documentation

[Contributing Guide](https://ucc.readthedocs.io/en/latest/contributing.html)  
[Setting up your Developer Environment](https://ucc.readthedocs.io/en/latest/contributing.html#setting-up-your-development-environment)  
[Writing a Custom Transpiler Pass](https://ucc.readthedocs.io/en/latest/user_guide.html#writing-a-custom-pass) 
-We use "transpiler pass" to refer to transformations that act only on the Directed Acyclic Graph (DAG) representation of the quantum circuit. Higher or lower-level optimizations (e.g. algorithm-level or pulse-level, respectively), we call "compiler passes."

**Comment by jordandsullivan** [comment]
Thanks for your proposal @poig ! It sounds interesting. I'm a bit confused that UCC has a uniform predicted fidelity of 0.0 in your first plot. Is your data normalized so that UCC is set to zero, or is there something I'm missing?

**Reply by poig** [reply]
> Thanks for your proposal @poig ! It sounds interesting. I'm a bit confused that UCC has a uniform predicted fidelity of 0.0 in your first plot. Is your data normalized so that UCC is set to zero, or is there something I'm missing?

because there are no layout for fake device with ucc, so I can't calculate the correct fidelity for it, where my fidelity calculation are based on the layout error rate

**Reply by poig** [reply]
and the reason you see the reliable going down so quick is because the train data(1000 random circuit) limit fewer toward that number of qubit circuit so transformer don't know what to predict the circuit token and 3minute trained small model just for demonstration the potential

**Reply by Misty-W** [reply]
We are adding layout benchmarks in the UCC benchmarking repo https://github.com/unitaryfoundation/ucc-bench/pull/78 to streamline the testing process for layout passes.
In the meantime, you can try importing one of the fake devices from Qiskit and using it to test. Fake devices generally have error models already built in.

**Reply by jordandsullivan** [reply]
@poig We merged in the new layout benchmarks in ucc-bench, so it should be easier for you to test out your new compiler pass :)

**Reply by poig** [reply]
done, just train some new model with kaggle free 2*T4 GPU within few minute

**Reply by jordandsullivan** [reply]
Okay @poig , I still don't see a PR from you in this repo, which is required if you want to be eligible for the bounty. Please submit one with the relevant commits so we can review :)

**Reply by poig** [reply]
@jordandsullivan thanks for reminder, just create a pr for it, but there are file I created just for training so everything still need to sort out
